File: systemdesign/SnakeAndLadder.py
import random

class Player:
    def __init__(self, id, name) -> None:
        self.id = id
        self.name = name

class Dice:
    def rollDice(self):
        return random.randint(1,10)

class Jumper:
    def __init__(self, startPoint, endPoint) -> None:
        self.startPoint = startPoint
        self.endPoint = endPoint
    
    def isLadder(self):
        return self.startPoint < self.endPoint
    
    def isSnake(self):
        return self.startPoint > self.endPoint

class GameBoard:
    def __init__(self, boardSize) -> None:
        self.boardSize = boardSize
        self.dice = Dice()
        self.nextTurn = list()
        self.snakes = list()
        self.ladders = list()
        self.playersCurrentPosition = dict()
    
    def play(self):
        while len(self.nextTurn) > 1:
            player = self.nextTurn.pop(0)
            currentPosition = self.playersCurrentPosition(player.name)
            diceValue = self.dice.rollDice()
            nextCell = currentPosition + diceValue
            if nextCell > self.boardSize:
                self.nextTurn.append(player)
            elif nextCell == self.boardSize:
                print (player.name + "has won the game")
            else:
                nextPosition = nextCell
                for snake in self.snakes:
                    if snake.startPoint == nextPosition:
                        nextPosition = snake.endPoint
                if nextPosition != nextCell:
                    print (player.name + "Bitten by Snake at position : " + nextCell)
                playerGotLadder = False
                for ladder in self.ladders:
                    if ladder.startPoint == nextPosition:
                        nextPosition = ladder.endPoint
                        playerGotLadder = True
                if nextPosition != nextCell and playerGotLadder:
                    print (player.name + "Got Ladder at position : " + nextCell)
                if nextPosition == self.boardSize:
                    print (player.name + "has won the game")
                else:
                    self.playersCurrentPosition[player.name, nextPosition]
                    print (player.name + "is at position : " + nextPosition)
                    self.nextTurn.append(player)
                

File: systemdesign/SnakeAndLadder.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

const (
	boardSize    = 100
	numSnakes    = 5
	numLadders   = 5
	snakePenalty = 5
	ladderBonus  = 10
)

var (
	snakes  = map[int]int{}
	ladders = map[int]int{}
)

func init() {
	rand.Seed(time.Now().UnixNano())
	generateSnakesAndLadders()
}

func generateSnakesAndLadders() {
	for i := 0; i < numSnakes; i++ {
		head := rand.Intn(boardSize-10) + 10
		tail := rand.Intn(head-1) + 1
		snakes[head] = tail
	}
	for i := 0; i < numLadders; i++ {
		bottom := rand.Intn(boardSize-10) + 1
		top := rand.Intn(boardSize-bottom) + bottom + 1
		ladders[bottom] = top
	}
}

func rollDice() int {
	return rand.Intn(6) + 1
}

func main() {
	var playerPosition int
	var currentPlayer int

	fmt.Println("Welcome to Snake and Ladder!")

	for playerPosition < boardSize {
		currentPlayer = (currentPlayer + 1) % 2

		fmt.Printf("Player %d's turn (Press Enter to roll the dice)\n", currentPlayer+1)
		fmt.Scanln()

		diceRoll := rollDice()
		fmt.Printf("Player %d rolled a %d\n", currentPlayer+1, diceRoll)

		playerPosition += diceRoll
		if newPosition, ok := snakes[playerPosition]; ok {
			fmt.Printf("Player %d encountered a snake! Moving to position %d\n", currentPlayer+1, newPosition)
			playerPosition = newPosition
		} else if newPosition, ok := ladders[playerPosition]; ok {
			fmt.Printf("Player %d found a ladder! Moving to position %d\n", currentPlayer+1, newPosition)
			playerPosition = newPosition
		}

		if playerPosition > boardSize {
			playerPosition = boardSize - (playerPosition - boardSize)
		}

		fmt.Printf("Player %d is now at position %d\n", currentPlayer+1, playerPosition)

		if playerPosition == boardSize {
			fmt.Printf("Player %d wins!\n", currentPlayer+1)
			break
		}
	}
}
